dev/com.ibm.ws.config/src/com/ibm/ws/config/admin/internal/ConfigurationStorageHelper.java

```diff
@@ -0,0 +1,638 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.ibm.ws.config.admin.ConfigID;
+import com.ibm.ws.config.admin.ConfigurationDictionary;
+import com.ibm.wsspi.kernel.service.utils.OnErrorUtil.OnError;
+import com.ibm.wsspi.kernel.service.utils.SerializableProtectedString;
+
+/**
+ * A helper class for storing and retrieving the persisted config information.
+ *
+ * The format is:
+ *
+ * byte - version, set to 0
+ * utf - bundleLocation
+ * int - number of unique variables
+ * utf - unique variables, write the number of entries dictated by prior
+ * int - number of config references
+ * configIDs - the config ids written parent to child in order.
+ * int - optionaly. Written if this a parent to indicate how many entries in this tree
+ * utf - the id
+ * utf - the pid
+ * utf - the child attribute
+ * int - an index to which config id is the parent. If an entry is -1 it means it has no parent.
+ * int - the number of entries in the config dictionary
+ * entries - the entries in the config dictionary
+ * utf - the name of the entry
+ * int - the type of the entry
+ * <value> - the value. How this is written depends on whether it is primitive, collection, array or map.
+ * - primitives are written directly
+ * - collections have their size written followed by each entry
+ * - arrays have their size written followed by each entry
+ * - maps are written the same way as the main config dictionary
+ */
+public class ConfigurationStorageHelper {
+
+    private static final int VERSION = 0;
+    private static final int BYTE = 0;
+    private static final int SHORT = 1;
+    private static final int BOOLEAN = 2;
+    private static final int CHAR = 3;
+    private static final int DOUBLE = 4;
+    private static final int LONG = 5;
+    private static final int FLOAT = 6;
+    private static final int INTEGER = 7;
+    private static final int STRING = 8;
+    private static final int PROTECTED_STRING = 9;
+    private static final int COLLECTION = 10;
+    private static final int MAP = 11;
+    private static final int ARRAY = 12;
+    private static final int ONERROR = 13;
+
+    public static String load(File configFile, Set<String> uniqueVars, Set<ConfigID> references, ConfigurationDictionary dict) throws IOException {
+        try (DataInputStream dos = new DataInputStream(new BufferedInputStream(new FileInputStream(configFile)))) {
+
+            if (dos.readByte() == VERSION) {
+                String location = null;
+                if (dos.readBoolean()) {
+                    location = dos.readUTF();
+                }
+                int len = dos.readInt();
+                for (int i = 0; i < len; i++) {
+                    uniqueVars.add(dos.readUTF());
+                }
+                len = dos.readInt();
+                List<ConfigID> ids = new ArrayList<>();
+                for (int i = 0, j = 0, cursor = 0; i < len + cursor; i++, j++) {
+                    if (cursor == 0) {
```

Comment:
The `cursor` variable name is confusing me.  I think this relates to the `depth` variable in `writeConfigID` or perhaps `j` is more like the `depth` in read and `cursor` is the `floor` you are trying to reach?